### PR TITLE
Sonnet V1.2.5

### PR DESCRIPTION
**Pull request by ultrabear**
Good enough for a release ig
whatever

### CODE REVIEW

File: cmds/cmd_utils.py [hunk 1]
```diff
@@ -146,35 +142,43 @@ async def help_function(message: discord.Message, args: List[str], client: disco
...
+            if "rich_description" in cmds_dict[a]:
+                cmd_embed.add_field(name="Detailed information:", value=cmds_dict[a]["rich_description"], inline=False)
 
-            if (t := type(kwargs["cmds_dict"][a]["permission"])) == str:
-                perms = kwargs["cmds_dict"][a]["permission"]
+            if (t := type(cmds_dict[a]["permission"])) == str:
+                perms = cmds_dict[a]["permission"]
             elif t == tuple or t == list:
-                perms = kwargs["cmds_dict"][a]["permission"][0]
+                perms = cmds_dict[a]["permission"][0]
```
Comment (ultrabear) [inline cmds/cmd_utils.py:167]: oops ill fix this next release (its not broken dw)

**Comment by ultrabear** [comment]
# Changelog
## Frontend:
- Add: error handling for GDPR download files too big to upload
- Add: leave-log and message-edit-log overloads to logging channels, allows better seperation
- Add: joinrules command subsets, allows user, timestamp, and default pfp warnings
- Add: Show who the bot could not dm in infraction logging
- Add: error parsing for forbidden errors on sending embeds
- Add: search-infractions new flag for infractions per page
- Change: sonnetsh/map now tell how long it took to execute and print a sucess message
- Change: profile/pfp now support parsing members not in guild
- Change: help command now shows total commands and total endpoints
- Change: serverinfo now shows top role as a mention
- Change: deletion log timestamps now show message creation time, not deletion time
- Change: parse_log_channel and parse_role now allow typing remove|rm|delete to remove the log item
## Under the Hood:
- Add: more typing to build_tools
- Add: more typing to sonnet core and everywhere
- Change: Seperate files from discord.File in GDPR command
- Add: constants library for discord limits and sonnet error strings
- Add: delete_config db_hlapi endpoint
- Add: db_hlapi lock kwarg to allow for locking database queries per guild
- Add: lib_parsers.parse_user_member to parse a user and member object from args
- Change: search-infractions now uses SQL to query for infractions
- Change: search-infractions now uses math to calculate page counts
- Change: sonnetsh/map now use try: finally: syntax to gaurantee the message.content is not lost
- Add: limits to deletion and editing logs to not go over (cause 4k char limits)
- Add: dummy typing class for cryptography
- Add: cmd_module, dlib_module and _dict counterparts to lib_lexdpyk_h
- Add: lib_loaders endpoint to get_guild_lock for threadlocks
- Add: TEXT_KEY endpoint to db_handler that says if TEXT is allowed to be used as a key or not
- Add: make_new_index endpoint to db_handler to allow for creating indexes
- Add: multifetch_rows_from_table endpoint to db_handler that allows fetching based on multiple datapoints
- Add: multicount_rows_from_table endpoint to db_handler that allows fetching a count of the db query instead of the data itself
- Add: ping endpoint to db_handler to ping if db connection is alive
- Add: Deprecationwarning on lib_parsers.ifgate
- Add: docstrings to all lib_parsers endpoints
- Add: LeXdPyK V1.3.3:
  - Add: typing to all of kernel
  - Add: error logging to err.log
  - Add: errors are not raised to dpy, instead printed with traceback by the kernel
  - Add: handle discord.errors.Forbidden on sending error messages from kernel
- Change: grab_files returns bare io.BytesIO discord.File to allow for no underlying circular references
- Add: grab_filter_infractions to db_hlapi, deprecates grab_user_infractions and grab_moderator_infractions
## Runtime Environment:
- Add: autotest easy threading test script
- Change: bump requirements

File: libs/lib_sonnetdb.py [hunk 2]
```diff
@@ -310,7 +358,9 @@ def is_muted(self, userid: Optional[int] = None, infractionid: Optional[str] = N
     def close(self) -> None:
         self.database.commit()
 
-    def __exit__(self, err_type, err_value, err_traceback):
+    def __exit__(self, err_type: Optional[Type[Exception]], err_value: Optional[str], err_traceback: Any) -> None:
+        if self._lock:
+            self._lock.release()
```
Comment (Saragon4005) [inline libs/lib_sonnetdb.py:363]: wouldn't you want to commit first?

File: libs/lib_sql_handler.py [hunk 3]
```diff
@@ -84,15 +101,50 @@ def add_to_table(self, table: str, data: Union[List[Any], Tuple[Any, ...]]) -> N
...
+        # Execute
+        self.cur.execute(db_inputStr, tuple(db_inputList))
+
+        retval: int = tuple(self.cur.fetchall())[0][0]
+        return retval
+
+    def fetch_rows_from_table(self, table: str, search: List[Any]) -> Tuple[Any, ...]:
+
+        # Test for attack
+        if table.count("\\") or table.count("'"):
```
Comment (Saragon4005) [inline libs/lib_sql_handler.py:125]: you used `in` previously

File: libs/lib_sql_handler.py [hunk 4]
```diff
@@ -84,15 +101,50 @@ def add_to_table(self, table: str, data: Union[List[Any], Tuple[Any, ...]]) -> N
...
+
+        # Execute
+        self.cur.execute(db_inputStr, tuple(db_inputList))
+
+        return tuple(self.cur.fetchall())
+
+    def multifetch_rows_from_table(self, table: str, searchparms: List[List[Any]]) -> Tuple[Any, ...]:
+
+        # Test for attack
+        if table.count("\\") or table.count("'"):
```
Comment (Saragon4005) [inline libs/lib_sql_handler.py:140]: here as well

File: main.py [hunk 5]
```diff
@@ -436,19 +446,26 @@ def logging_toggle(*args):
...
+        with open("err.log", "a+") as logfile:
+            logfile.write(f"AT {time.strftime('%a, %d %b %Y %H:%M:%S', datetime.datetime.utcnow().utctimetuple())}:\n")
+            logfile.write("".join(traceback.format_exception(type(self.err), self.err, self.err.__traceback__)))
 
-# Catch errors.
+
+# Catch errors
 @Client.event
 async def on_error(event, *args, **kwargs):
     raise
```
Comment (Saragon4005) [inline main.py:465]: I am 90% sure it does this on its own

**Review by Saragon4005 — CHANGES_REQUESTED** [review]

File: main.py (hunk 5 above)
Comment (ultrabear) [reply main.py:465]: well removing it makes it crash on error so clearly not

File: libs/lib_sonnetdb.py (hunk 2 above)
Comment (ultrabear) [reply libs/lib_sonnetdb.py:363]: what if the commit errors
mm deadlock

File: cmds/cmd_utils.py [hunk 6]
```diff
@@ -17,37 +17,20 @@
...
+importlib.reload(lib_constants)
+import lib_lexdpyk_h
+
+importlib.reload(lib_lexdpyk_h)
 
 from lib_db_obfuscator import db_hlapi
-from lib_parsers import parse_permissions, parse_boolean
+from lib_parsers import parse_permissions, parse_boolean, parse_user_member
 from lib_loaders import load_embed_color, embed_colors
+import lib_constants as constants
```
[the remark this answers is not in the document]
Comment (ultrabear) [reply cmds/cmd_utils.py:30]: importlib.reload does things to you
including that annoyance

File: cmds/cmd_scripting.py [hunk 7]
```diff
@@ -5,110 +5,128 @@
...
+                    # Stop processing if error
+                    if suc != 0:
+                        await message.channel.send(f"ERROR: {self_name}: command `{command}` exited with non success status")
+                        return 1
+
+                    # Regenerate cache
+                    if cmds_dict[command]['cache'] in ["purge", "regenerate"]:
+                        cache_purge = True
+                else:
+                    # dont forget to re reference message content even if exec stops
```
[the remark this answers is not in the document]
Comment (ultrabear) [reply cmds/cmd_scripting.py:103]: you spoon finally runs every single time no matter what happens, including returns, nonexeptions, and exceptions

File: cmds/cmd_automod.py [hunk 8]
```diff
@@ -320,9 +324,161 @@ async def antispam_time_set(message: discord.Message, args: List[str], client: d
...
+            await self.m.channel.send(f"new user notify is set to {jointime} seconds")
+            return 0
+
+    async def defaultpfpedit(self, args: List[str], client: discord.Client) -> int:
+        # notifier-log-defaultpfp
+        cnf_name: Final[str] = "notifier-log-defaultpfp"
+
+        if args:
+            with db_hlapi(self.m.guild.id) as db:
+                db.add_config(cnf_name, str(true := int(parse_boolean(args[0]))))
```
[the remark this answers is not in the document]
Comment (ultrabear) [reply cmds/cmd_automod.py:446]: its used once it doesnt matter that much

File: cmds/cmd_automod.py [hunk 9]
```diff
@@ -320,9 +324,161 @@ async def antispam_time_set(message: discord.Message, args: List[str], client: d
...
+        nsv: List[str] = [f"{i} {self.ops[i][1]}\n" for i in self.ops]
+        await self.m.channel.send(f"JoinRule Help```py\n{''.join(nsv)}```")
+        return 0
+
+    async def useredit(self, args: List[str], client: discord.Client) -> int:
+        # notifier-log-users
+        cnf_name: Final[str] = "notifier-log-users"
+
+        if len(args) >= 2:
+            if args[0] == "add":
```
[the remark this answers is not in the document]
Comment (ultrabear) [reply cmds/cmd_automod.py:349]: this is too abstract to try and solve

File: libs/lib_sql_handler.py (hunk 3 above)
Comment (ultrabear) [reply libs/lib_sql_handler.py:125]: its the difference between str.__contains__ and str.count its not that much slower i cba to rewrite the rest

**Review by Saragon4005 — APPROVED** [review]

**Pull request merged**